java/src/main/native/OrtJniUtil.c

```diff
@@ -600,427 +588,473 @@ size_t copyPrimitiveArrayToJava(JNIEnv *jniEnv, ONNXTensorElementDataType onnxTy
+    if (code != ORT_OK) {
+        return code;
+    }
 
     // Create the character and offset buffers
-    char * characterBuffer;
-    checkOrtStatus(jniEnv,api,api->AllocatorAlloc(allocator,sizeof(char)*(totalStringLength+length),(void**)&characterBuffer));
+    char * characterBuffer = NULL;
+    code = checkOrtStatus(jniEnv, api, api->AllocatorAlloc(allocator, sizeof(char)*(totalStringLength+length), (void**)&characterBuffer));
+    if (code != ORT_OK) {
+        return code;
+    }
     // length + 1 as we need to write out the final offset
-    size_t * offsets;
-    checkOrtStatus(jniEnv,api,api->AllocatorAlloc(allocator,sizeof(size_t)*(length+1),(void**)&offsets));
+    size_t * offsets = NULL;
+    code = checkOrtStatus(jniEnv, api, api->AllocatorAlloc(allocator, sizeof(size_t)*(length+1), (void**)&offsets));
+    if (code != ORT_OK) {
+        checkOrtStatus(jniEnv,api,api->AllocatorFree(allocator,characterBuffer));
+        return code;
+    }
 
     // Get a view on the String data
-    checkOrtStatus(jniEnv,api,api->GetStringTensorContent(tensor,characterBuffer,totalStringLength,offsets,length));
-
-    // Get the final offset, write to the end of the array.
-    checkOrtStatus(jniEnv,api,api->GetStringTensorDataLength(tensor,offsets+length));
-
-    char * tempBuffer = NULL;
-    size_t bufferSize = 0;
-    for (size_t i = 0; i < length; i++) {
-        size_t curSize = (offsets[i+1] - offsets[i]) + 1;
-        if (curSize > bufferSize) {
-            checkOrtStatus(jniEnv,api,api->AllocatorFree(allocator,tempBuffer));
-            checkOrtStatus(jniEnv,api,api->AllocatorAlloc(allocator,curSize,(void**)&tempBuffer));
-            bufferSize = curSize;
-        }
-        if (tempBuffer == NULL) {
-            throwOrtException(jniEnv, 1, "Not enough memory");
-            return;
+    code = checkOrtStatus(jniEnv, api, api->GetStringTensorContent(tensor, characterBuffer, totalStringLength, offsets, length));
+    if (code == ORT_OK) {
+        // Get the final offset, write to the end of the array.
+        code = checkOrtStatus(jniEnv, api, api->GetStringTensorDataLength(tensor, offsets+length));
+        if (code == ORT_OK) {
+            size_t bufferSize = 0;
+            for (size_t i = 0; i < length; i++) {
+                size_t curSize = (offsets[i+1] - offsets[i]) + 1;
+                if (curSize > bufferSize) {
+                    code = checkOrtStatus(jniEnv, api, api->AllocatorFree(allocator, tempBuffer));
+                    if (code != ORT_OK) {
+                        goto string_tensor_cleanup;
+                    }
+                    code = checkOrtStatus(jniEnv, api, api->AllocatorAlloc(allocator, curSize, (void**)&tempBuffer));
+                    bufferSize = curSize;
+                }
+                if (tempBuffer == NULL) {
+                    throwOrtException(jniEnv, 1, "Not enough memory");
```

Comment:
This should have been already checked/reported by checkOrtStatus above? Also tempBuffer is  allocated only conditionally.
Should we move this check closer to line 700.

Reply:
I rearranged it. It's a little messy as the first allocation is unconditional but we don't know how big it is until we get into the loop so it looks conditional. It looks a bit cleaner now.